amacc.c

```diff
@@ -747,63 +765,62 @@ void stmt()
      */
     case While:
         next();
-        a = e + 1; // start address of "while" body
         if (tk == '(') next();
         else fatal("open paren expected");
-        expr(Assign);
+        expr(Assign); a = n;
         if (tk == ')') next();
         else fatal("close paren expected");
-        *++e = BZ; b = ++e;
         stmt(); // parse body of "while"
-        /* unconditional jump to the start of "while" statement
-         * (including the code for the loop condition), to implement the loop
-         */
-        *++e = JMP; *++e = (int) a;
-        *b = (int) (e + 1); // "BZ" jump target (end of cycle)
+        *--n = (int) a; *--n = While;
         return;
     case Switch:
+        i = 0; j = 0;
+        if(cas) j = (int)cas; 
+        cas = &i;
         next();
         if (tk == '(') next();
         else fatal("open paren expected");
-        expr(Assign);
+        expr(Assign); 
+        a = n;
         if (tk == ')') next();
         else fatal("close paren expected");
-        a = cas; *++e = JMP; cas = ++e;
-        b = brks; d = def; brks = def = 0;
         stmt();
-        *cas = def ? (int) def : (int) (e + 1); cas = a;
-        while (brks) { a = (int *) *brks; *brks = (int) (e + 1); brks = a; }
-        brks = b; def = d;
+        b = n;
+        *--n = (int) b; *--n = (int) a; *--n = Switch;
+        if(j) cas = (int*) j;
         return;
     case Case:
-        *++e = JMP; ++e;
-        *e = (int) (e + 7); *++e = PSH; i = *cas; *cas = (int) e;
+        i = *cas;
         next();
         expr(Or);
-        if (e[-1] != IMM) fatal("bad case immediate");
-        *e = *e - i; *++e = SUB; *++e = BNZ; cas = ++e; *e = i + e[-3];
+        a = n;
+        if (*n != Num) fatal("bad case immediate");
+        j = n[1]; n[1] -= i; *cas = j;
+        *--n = ';';
         if (tk == ':') next();
         else fatal("colon expected");
         stmt();
+        b = n;
+        *--n = (int) b;*--n = (int) a;*--n = Case;
         return;
     case Break:
         next();
         if (tk == ';') next();
         else fatal("semicolon expected");
-        *++e = JMP; *++e = (int) brks; brks = e;
+        *--n = Break;
         return;
     case Default:
         next();
         if (tk == ':') next();
         else fatal("colon expected");
-        def = e + 1;
-        stmt();
+        stmt(); a = n;
+        *--n = (int) a; *--n = Default;
         return;
     // RETURN_stmt -> 'return' expr ';' | 'return' ';'
     case Return:
-        next();
-        if (tk != ';') expr(Assign); // calculate the return value
-        *++e = LEV;
+        a = 0; next();
+        if (tk != ';'){ expr(Assign); a = n; }
```

Comment:
Move the comment back.

Reply:
The motivation is different inside expr(),so I remove it.